Feature/sync GitHub profiles

ynang opened the pull request.

ynang committed:
代码同步

代码同步

ynang committed:
代码命名规范和格式优化

代码命名规范和格式优化

ynang committed:
规范代码和优化代码结构和思路

1、规范简化代码变量名和方法名；
2、简化条件判断内容，在通用情况下，由if val is not None改为if not val；
3、填补逻辑漏洞：对集合严格判空、对变量先判断后引用；
4、将通用方法全部集中于util包中；
5、删除应该位于后续与clickhouse对接中的对profile的统一清洗；
6、将同步的profile集合中的check node判断交与os中的query处理；
7、简化同步profile的时间范围判断代码；
8、优化同步profile代码的整体逻辑。

ynang committed:
优化同步profile逻辑

优化同步profile逻辑

ynang committed:
同步profile逻辑优化

同步profile逻辑优化

ynang committed:
同步profile逻辑优化

1、collapse和scan、scroll无法同时使用，为保证效率， 将每次获取的数据量设定为1000；
2、一小时只查询比较不存储约处理360个profile,为保证高效可用，每次查询时间由2小时延长至3小时；
3、优化每次循环结束时check node的存储，从查询表中id最小profile更改为本次1000profile查询更新结束后剩余profile表中需要
更新的id最小值

ynang committed:
同步profile逻辑优化

1、将循环查询更新中的判断是否在查询有效时间置于查询更新之后；
2、简化插入check node为一次

crystaldust committed:
Test the profile sync

ynang committed:
对查询更新GitHub profile模块的逻辑重构、技术栈转换、代码编写

1、为解决使用os存储更新的效率低下和以os为数据源airflow实现线程并发以及判断的复杂逻辑，引入redis作为查询更新profile过程中的缓存数据库；
2、将从os中获取profile列表方式从使用折叠方式的10000条限制转换为用scan获取；
3、根据以上两条进行GitHub profile模块的逻辑重构和代码编写。
Merge remote-tracking branch 'meng/main' into main

ynang committed:
对查询更新GitHub profile模块的逻辑重构、技术栈转换、代码编写

1、为解决使用os存储更新的效率低下和以os为数据源airflow实现线程并发以及判断的复杂逻辑，引入redis作为查询更新profile过程中的缓存数据库；
2、将从os中获取profile列表方式从使用折叠方式的10000条限制转换为用scan获取；
3、根据以上两条进行GitHub profile模块的逻辑重构和代码编写。

ynang committed:
完善查询更新profile逻辑处理

1、增加根据id查询更新profile的task中开始的set判空；
2、增加根据id查询更新profile的task中查询更新过程中产生的异常处理。

ynang committed:
查询更新profile模块逻辑优化及规范方法和变量命名

1、对部分判断进行逻辑优化；
2、部分变量名常量化；
3、删除redis分布式锁；
4、redispy中的pipeline对于事务支持的变量类型未解决。
Merge remote-tracking branch 'meng/main' into main

ynang committed:
查询更新profile模块代码优化

1、配置类常量名置于airflow取用，其他常量置于本项目常量模块；
2、简化部分变量名；
3、对初始化以及查询profile模块的相应os中的index的结构进行同步调整，增加search_key以及raw_data；
4、增加查询更新profile模块的查询scan的scroll时间为30分钟；
5、对查询更新profile模块的redis的字典之间的数据传递由python转换为lua脚本，保证整个过程的原子性。

ynang committed:
查询更新profile模块代码优化

1、配置类常量名置于airflow取用，其他常量置于本项目常量模块；
2、简化部分变量名；
3、对初始化以及查询profile模块的相应os中的index的结构进行同步调整，增加search_key以及raw_data；
4、增加查询更新profile模块的查询scan的scroll时间为30分钟；
5、对查询更新profile模块的redis的字典之间的数据传递由python转换为lua脚本，保证整个过程的原子性。

Merge remote-tracking branch 'meng/main' into main

ynang committed:
查询更新profile模块代码优化

1、配置类常量名置于airflow取用，其他常量置于本项目常量模块；
2、简化部分变量名；
3、对初始化以及查询profile模块的相应os中的index的结构进行同步调整，增加search_key以及raw_data；
4、对初始化以及查询profile模块的异常处理统一profile结构；
5、增加查询更新profile模块的查询scan的scroll时间为30分钟；
6、对查询更新profile模块的redis的字典之间的数据传递由python转换为lua脚本，保证整个过程的原子性。

ynang committed:
查询更新profile模块代码优化

1、配置类常量名置于airflow取用，其他常量置于本项目常量模块；
2、简化部分变量名；
3、对初始化以及查询profile模块的相应os中的index的结构进行同步调整，增加search_key以及raw_data；
4、对初始化以及查询profile模块的异常处理统一profile结构；
5、增加查询更新profile模块的查询scan的scroll时间为30分钟；
6、对查询更新profile模块的redis的字典之间的数据传递由python转换为lua脚本，保证整个过程的原子性。

ynang committed:
查询更新profile模块代码优化

1、配置类常量名置于airflow取用，其他常量置于本项目常量模块；
2、简化部分变量名；
3、对初始化以及查询profile模块的相应os中的index的结构进行同步调整，增加search_key以及raw_data；
4、对初始化以及查询profile模块的异常处理统一profile结构；
5、增加查询更新profile模块的查询scan的scroll时间为30分钟；
6、对查询更新profile模块的redis的字典之间的数据传递由python转换为lua脚本，保证整个过程的原子性。Merge remote-tracking branch 'oss-know/development' into main

crystaldust commented on dags/oss_know/libs/github/sync_profiles.py:
```diff
@@ -1,198 +1,115 @@
+from opensearchpy import helpers as opensearch_helpers
+
+
+def get_redis_client(redis_client_infos):
```
Consider moving the function to utils or base under lib, this is a pretty common function, should be provided to all modules, not only sync profile module.

crystaldust commented on dags/oss_know/libs/github/sync_profiles.py:
```diff
@@ -1,198 +1,115 @@
+    '''Put GitHub profiles that need to be updated into dict named sync_github_profiles:storage_updated_hash of redis.'''
+    opensearch_client = get_opensearch_client(opensearch_conn_info)
+    redis_client = get_redis_client(redis_client_info)
+    working_updated_hash=redis_client.hgetall(WORKING_HASH)
```
Always mind the coding style, using code formatter before pushing.

crystaldust commented on dags/oss_know/libs/github/sync_profiles.py:
```diff
@@ -1,198 +1,115 @@
+        redis_client.delete(WORKING_HASH)
+
+
+
```
Generally 2 newlines are enough according to the PEP standard.

crystaldust commented on dags/oss_know/libs/github/sync_profiles.py:
```diff
@@ -1,198 +1,115 @@
+    opensearch_client = get_opensearch_client(opensearch_conn_info)
+    redis_client = get_redis_client(redis_client_info)
+    # end_time = (datetime.datetime.now() + datetime.timedelta(seconds=duration_of_sync_github_profiles["seconds"])).timestamp()
+    end_time = (datetime.datetime.now() + datetime.timedelta(milliseconds=5000)).timestamp()
```
[High priority!] Read the duration from an airflow variable and pass it to the function. Always remove the uncenessary comments and testing code snippets before making a PR. PRs should be taken seriously!

crystaldust commented on dags/oss_know/libs/github/sync_profiles.py:
```diff
@@ -1,198 +1,115 @@
-
-
+                logger.info(f"{storage_id}'s github profiles of opensearch is latest.")
+        except Exception as e:
+            logger.error(f"Failed sync_github_profiles,the exception message: {e}")
+        else:
+            redis_client.hdel(WORKING_HASH, storage_id)
+        finally:
+            if datetime.datetime.now().timestamp() > end_time:
+                logger.info('The connection has timed out.')
+                break
```
This is a great trick to take python's try-except mechanism [thumb up]

crystaldust commented on dags/oss_know/libs/util/github_api.py:
```diff
@@ -51,8 +51,12 @@ def get_github_profiles(self, http_session, github_tokens_iter, id_info):
             logger.error(f"捕获airflow抛出的TypeError:{e}")
-            return {'id': id_info, 'status': 'User status is abnormal.', 'company': '', 'location': '',
-                    'email':''}
+            return {'login': '', 'id': id_info, 'node_id': '', 'avatar_url': '', 'gravatar_id': '', 'url': '', 'html_url': '',
```
- Make the code simpler with python's try-except mechanism
- If the function is to get user's latest profile, the function should be named 'get_latest_github_profile', then with the param user_id, we know the function is to get a github user's latest profile, parameter is descriptive.
- id_info is a bad var name, making developers think it's an object that contains lots of infos, here it's just an user_id

crystaldust commented on data_schema/sync_data.json:
```diff
@@ -26,6 +26,10 @@
         "sync_since_datetime": "2021-01-01T00:00:00+00:00",
         "sync_until_datetime": "2021-01-01T00:00:00+00:00"
+      },
+      "profiles": {
```
Remove if it's not necessary

ynang committed:
完善同步更新GItHub profile功能代码

1、规范代码格式及命名；
2、删除测试代码及非必要代码块。
Merge remote-tracking branch 'meng/main' into sync-profile

crystaldust commented on dags/oss_know/libs/util/github_api.py:
```diff
@@ -39,26 +39,27 @@ def get_github_issues(self, http_session, github_tokens_iter, owner, repo, page,
         """Get GitHub user's latest profile from GitHUb."""
         url = "https://api.github.com/user/{id_info}".format(
-            id_info=id_info)
+            id_info=user_id)
```
Use f string for clarity

ynang committed:
统一字符串格式为f"...{引用值}"

统一字符串格式为f"...{引用值}"
